### PR TITLE
[ADP-1963] Break up API types into submodules

### PR DESCRIPTION
**Pull request by paolino**
I have
- [x] factor out `ApiDecodedTransaction` from the Api.Types module
- [x] created ApiT, ApiAsArray, ExtendedObject, Options as API  libraries
- [x] factor out `Certificate` and `Key` API types with relatives closures
- [x] factor out `Primitive` for all ApiT instances that just wrap types in the `Primitive` hierarchy
- [x] moved out a couple of coin functions to the `Coin` primitive

### Comments

This work is preliminary to ApiDecodedTransaction reuse in the ADP work, but independent from CBOR PRs

### Issue Number

ADP-1963

### CODE REVIEW

File: lib/core/src/Cardano/Wallet/Primitive/Types/Coin.hs [hunk 1]
```diff
@@ -160,6 +162,11 @@ toQuantity (Coin c) = Quantity <$> intCastMaybe c
 toWord64 :: Coin -> Maybe Word64
 toWord64 (Coin c) = intCastMaybe c
 
+coinToQuantity :: Integral n => Coin -> Quantity "lovelace" n
+coinToQuantity = Quantity . Prelude.fromIntegral . unCoin
+
+coinFromQuantity :: Integral n => Quantity "lovelace" n -> Coin
+coinFromQuantity = Coin . Prelude.fromIntegral . getQuantity
```
Comment (jonathanknowles) [inline lib/core/src/Cardano/Wallet/Primitive/Types/Coin.hs:192]: Hi @paolino

I think we should get rid of these unsafe `fromIntegral` conversions where possible.

I've made a PR to consolidate all the various functions that convert between `Coin` and `Quantity` here: https://github.com/input-output-hk/cardano-wallet/pull/3492

Comment (paolino) [reply lib/core/src/Cardano/Wallet/Primitive/Types/Coin.hs:192]: should I remove these functions ?

**Comment by paolino** [comment]
CI failed:
``` 
To rerun use: --match "/Cardano.Wallet.Api.Server.Tls/TLS Client Authentication/Deny client with wrong certificate if TLS is enabled/"

Randomized with seed 1088953760
```
3 times

**Comment by paolino** [comment]
bors try

File: lib/core/src/Cardano/Wallet/Api/Types/Address.hs [hunk 2]
```diff
@@ -0,0 +1,48 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Cardano.Wallet.Api.Types.Address where
```
Comment (HeinrichApfelmus) [inline lib/core/src/Cardano/Wallet/Api/Types/Address.hs]:
```suggestion
module Cardano.Wallet.Api.Types.Address (
   …
   ) where
```

Could you add an explicit export list?

File: lib/core/src/Cardano/Wallet/Primitive/Types/Coin.hs (hunk 1 above)
Comment (HeinrichApfelmus) [reply lib/core/src/Cardano/Wallet/Primitive/Types/Coin.hs:192]: > should I remove these functions ?

Yes, please — remove them and instead import the functions that Jonathan added (they are in `master` now).

**Review by HeinrichApfelmus — APPROVED** [review]
Looks good to me, thanks! 😊 Requesting one explicit export list. Also, for the new modules, could you add the standard copyright notice on top:

```
-- |
-- Copyright: © 2018-2022 IOHK
-- License: Apache-2.0
module … where …
```

It's how we're supposed to do new modules. 🤓

**Review by jonathanknowles — APPROVED** [review]
Hi Paolo

This looks fine to me. I have the same suggestions as @HeinrichApfelmus  (regarding [exports](https://github.com/input-output-hk/cardano-wallet/pull/3491#discussion_r970990191) and [copyright notices](https://github.com/input-output-hk/cardano-wallet/pull/3491#pullrequestreview-1107765419)).

Would you mind if we were to change the title of the PR to "Break up API types into submodules" or something similar?

Since the scope of the PR is a bit larger than just factoring out the `ApiDecodedTransaction` type, I think it would help make the revision history more understandable if the merge commit title better reflected this larger scope!

**Pull request merged**